hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -856,31 +856,221 @@ def kill_branch(self) -> None:
         """Mark this part of the tree as not worth re-exploring."""
 
     def draw_integer(
-        self, value: int, *, was_forced: bool, kwargs: IntegerKWargs
+        self, value: int, *, kwargs: IntegerKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_float(
-        self, value: float, *, was_forced: bool, kwargs: FloatKWargs
+        self, value: float, *, kwargs: FloatKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_string(
-        self, value: str, *, was_forced: bool, kwargs: StringKWargs
+        self, value: str, *, kwargs: StringKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_bytes(
-        self, value: bytes, *, was_forced: bool, kwargs: BytesKWargs
+        self, value: bytes, *, kwargs: BytesKWargs, was_forced: bool
     ) -> None:
         pass
 
     def draw_boolean(
-        self, value: bool, *, was_forced: bool, kwargs: BooleanKWargs
+        self, value: bool, *, kwargs: BooleanKWargs, was_forced: bool
     ) -> None:
         pass
 
 
+IRTreeNodeType: TypeAlias = Union["IRTreeNode", "IRTreeLeaf"]
+LocationType: TypeAlias = Tuple[int, int]
+
+
+@attr.s(slots=True)
+class IRTreeLeaf:
+    ir_type: IRTypeName = attr.ib()
```

Comment:
I'd be tempted to set this in post-init, since it's trivially derivable from the `value`.

Reply:
I've been hesitant to derive `ir_type` from `value` here and elsewhere (`DataTree`), in anticipation of (very far in the future) potentially wanting to allow the ir layer to be extensible with new nodes that may overlap in types. I agree in principle it's weird to carry around this `ir_type` when all it currently encodes is `type(value)`.

Reply:
with the recent changes I'm not sure a derived value is cleaner than passing the ir type along, but feel free to make any changes you want here!